### ISSUE TITLE
[ Feature Request ] - Anaglyph 3D

### ISSUE THREAD

**Issue by ChristianSilvermoon** [Wohlstand/TheXTech#271]
When playing the 3DS build very early on, I'd noticed that @ds-sloth had gotten the 3D working.

I think it would be really nice to see [Anaglyph 3d](https://en.wikipedia.org/wiki/Anaglyph_3D) support so that the 3D effect seen in the 3DS build can be replicated on Desktop and other platforms.

**Comment by Wohlstand** [comment]
The fact, Hedgewars has that support. And yeah, I do have anaglyph glasses at me to test the thing in action 🦊
That means, will need to slightly tune the renderer and depending on the Z distance, draw texture with a small offset and color mod. (As at Hedgewars, there is also should be an ability to set up eye colors, because of different variants of glasses with different eye color.

P.S. Just a random useful stuff I found:
http://paulbourke.net/stereographics/anaglyph/

**Comment by Wohlstand** [comment]
P.S. there is some of Hedgewars side: https://github.com/hedgewars/hw/blob/master/hedgewars/uRender.pas
Find any mentions of `cStereoMode` where are conditions for the stereo rendering.

**Comment by ds-sloth** [comment]
Let's avoid adding Z distance to the engine. That will require us to maintain a sorted draw queue which will make drawing significantly more expensive. Additionally, I found that rendering each object twice (to get per-object offsets) significantly reduced performance.

In this case, we can implement 3D by giving the XRender a concept of "layers" where the different layers are different render targets (when 3D is enabled), and then they are combined at the end of the frame according to the 3D method we wish to use. You can see the `frmMain.setLayer` calls in the `devel` code.

I'm not familiar with SDL's render target system but I'm happy to try implementing this for SDL as well when I backport the feature from the `devel` branch.

**Comment by Wohlstand** [comment]
About Z-value per sprite (which is another topic, more general), that will be required in the future for scripts and organized render of objects (currently, it's a pain because of inconsistent BGO grouping and inability to flexibly control the Z-position of any in-game element). 

At Moondust Engine, I have the next thing here: https://github.com/WohlSoft/Moondust-Project/blob/master/Engine/scenes/level/lvl_camera.cpp
- All visible objects gets a place at the array of visible objects
- New-added elements will be appended into the list, and the list will be sorted
- Elements that go off the visible area will be popped without breaking the queue
- When no elements appeared/disappeared, an array of visible elements stay unchanged

I guess, for this, I'll make the new issue, it's more general than just adding an anaglyph effect.

**Comment by ds-sloth** [comment]
> that will be required in the future for scripts and organized render of objects

Yes, let's talk about this separately. It's possible that we can get hardware Z-values on most of our low-performance platforms since older platforms often have their own sprite engines, but I do not want a situation where I'm forced to sort a draw queue every frame on a low-performance platform.

However, I think the 3D effect is a separate question from draw order and they can be implemented separately.

**Comment by Wohlstand** [comment]
> Yes, let's talk about this separately. It's possible that we can get hardware Z-values on most of our low-performance platforms since older platforms often have their own sprite engines, but I do not want a situation where I'm forced to sort a draw queue every frame on a low-performance platform

I made the #272 issue with my explanations given at the main post.
